src/components/organisms/EntityDetails/EntityDetailsContent/Settings/SettingsTests/SettingsTests.tsx

```diff
@@ -74,162 +61,98 @@ const SettingsTests: React.FC<{openDefinition(): void}> = ({openDefinition}) =>
-    [entityDetails.steps]
-  );
+    return entityDetails.steps.map(step => {
+      return step.execute.map(item => {
+        const id = nanoid();
+        if ('delay' in item) {
+          return {
+            ...item,
+            id,
+          };
+        }
+
+        return {
+          ...item,
+          id,
+          type: testsData.find(x => x.name === item.test)?.type || '',
+        };
+      });
+    });
+  }, [testsData]);
 
-  const initialSteps: LocalStep[] = useMemo(
-    () =>
-      entityDetails.steps
-        ? entityDetails.steps.map(step => {
-            const id = nanoid();
+  const [isDelayModalVisible, setIsDelayModalVisible] = useState(false);
+  const [isTestModalVisible, setIsTestModalVisible] = useState(false);
+  const [currentGroup, setCurrentGroup] = useState<number | string>(0);
 
-            const firstItemInStep = step.execute[0];
+  const [steps, setSteps] = useState(initialSteps);
 
-            if ('delay' in firstItemInStep) {
-              return {
-                ...firstItemInStep,
-                id,
-              };
-            }
+  const wasTouched = steps !== initialSteps;
 
-            return {
-              ...firstItemInStep,
-              id,
-              type: testsData.find(item => item.name === firstItemInStep.test)?.type || '',
-              stopTestOnFailure: step.stopTestOnFailure,
-            };
-          })
-        : [],
-    [entityDetails?.steps, testsData]
-  );
+  useEffect(() => {
+    setSteps(initialSteps);
+  }, [testsData]);
 
-  const [currentSteps = initialSteps, setCurrentSteps] = useState<LocalStep[]>([]);
+  const showModal = (value: string, group: number | string) => {
+    if (value === 'test') {
+      setIsTestModalVisible(true);
+    } else {
+      setIsDelayModalVisible(true);
+    }
 
-  const wasTouched = currentSteps !== initialSteps;
+    setCurrentGroup(group);
+  };
 
-  useEffect(() => {
-    if (currentSteps !== initialSteps) {
-      setCurrentSteps(initialSteps);
+  const addNode = (data: number | {name: string; type: string}) => {
+    const item =
+      typeof data === 'number' ? {id: nanoid(), delay: `${data}ms`} : {id: nanoid(), test: data.name, type: data.type};
+
+    if (typeof currentGroup === 'number') {
+      setSteps([
+        ...steps.slice(0, currentGroup),
+        [...(steps[currentGroup] || []), item],
+        ...steps.slice(currentGroup + 1),
+      ]);
+    } else {
+      const index = Number(currentGroup.split('-')[0]) + 1;
+      setSteps([...steps.slice(0, index), [item], ...steps.slice(index)]);
     }
-  }, [initialSteps]);
+  };
+
+  const onSave = () => {
+    const resultSteps = steps.map((items, i) => {
+      return {
+        execute: items.map(item => {
+          if ('delay' in item) {
+            return {
+              delay: item.delay,
+            };
+          }
+          return {
+            test: item.test,
+          };
+        }),
```

Comment:
Optionally, that could use `lodash/pick` with `items.map(item => pick(item, ['delay', 'test']))`

Reply:
Oh, this is nice